### ISSUE TITLE
Support sparse matrices with more than 2^31-1 elements

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#3513]
Currently there are a number of places in the codebase where the dtype for `indptr` and `indices` are hardcoded to `'i'` (i.e., `int32`), for example: 
https://github.com/cupy/cupy/blob/22e270f303603848584974d1fbee5a38c9a5c428/cupyx/scipy/sparse/compressed.py#L330-L331
But they shouldn't be. In SciPy, they use a helper function `get_index_dtype()` to determine this:
https://github.com/scipy/scipy/blob/f4b5605031f738bc87ae4e193d614d525b98ffba/scipy/sparse/compressed.py#L65-L71
I think this change shouldn't be hard, but it is worth a standalone PR with sufficient test coverage added. 

For the short term, I'd say let's encourage PR authors to take this into account, so we can make the changes gradually. In particular, when a new `ElementwiseKernel` or `ReductionKernel` is added, we just need to use template types `T` as oppose to hardcoded `int32`.

**Comment by takagi** [comment]
It seems that the newer cuSPARSE Generic API supports 64bit indices (and the legacy API we currently use only supports 32bit indices).

https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-api-reference

**Comment by takagi** [comment]
cuSPARSE Generic API is starting with CUDA 10.1.
https://docs.nvidia.com/cuda/archive/10.1/cusparse/index.html#cusparse-generic-api-reference

**Comment by takagi** [comment]
CuPy already has cuSPARSE Generic API supports, #3129 and #3242, awesome @anaruse!

**Comment by leofang** [comment]
That's right, so using 64-bit indices becomes possible!

**Comment by leofang** [comment]
When working on #4778 I noticed one obstacle: In some index calculations we need `scatter_add` to handle `int64`, but it relies on CUDA's builtin `atomicAdd`, which does not have a `int64` specialization. Don't see an easy workaround ftm.

ref: https://github.com/cupy/cupy/pull/4778/commits/54227888221679ece62ec70273fcfdada281202d

**Comment by kmaehashi** [comment]
I noticed docs say several cuSPARSE APIs will be removed in the next major release. e.g.

https://docs.nvidia.com/cuda/cusparse/index.html#csrgemm2

```
11.2. cusparse<t>csrgemm2() [DEPRECATED]
[[DEPRECATED]] use cusparseSpGEMM() instead. The routine will be removed in the next major release
```

Maybe we should prioritize this?

**Comment by leofang** [comment]
Yes there are a ton of deprecation warnings emitted when compiling the cuSPARSE modules. Even without considering the 64-bit support it's still nice if we can remove these functions.

**Comment by takagi** [comment]
https://github.com/cupy/cupy/issues/3513#issuecomment-735208809 Fortunately, the oldest CTK that CuPy supports (10.2)  has the Generic API now.

**Comment by Intron7** [comment]
Is there still work being done on 64bit indptr support for future cupy versions?

**Comment by kmaehashi** [comment]
We haven't started working on this one.

@takagi We're getting requests for this one several times so maybe we need to reconsider the priority?

**Comment by takagi** [comment]
Agree. I've marked this as high priority and have will to work on this.

**Comment by tabedzki** [comment]
Hi there, 

We're using rapids-singlecell which relies on this package. We get similar errors to the ones from [scverse/rapids_singlecell#43](https://github.com/scverse/rapids_singlecell/issues/43) and [scverse/rapids_singlecell#131](https://github.com/scverse/rapids_singlecell/issues/131). 

Where does this stand/how can we help? We have some bandwidth at a Hackathon to help address this.

**Comment by leofang** [comment]
Assigning to myself so that I remember to find someone to help take over 🙂

**Comment by samaid** [comment]
I've been playing lately to see what it will take to support 64-bit indices for sparse matrix. My conclusion is that most of the issues can be addressed with existing CuPy functionality. However, there is one "evil" that needs additional research. Specifically, the use of random choice (permutation) to initialize sparse matrices randomly by density. The existing CuPy algorithm is the outdated one from Knuth book which is largely unsuitable for GPUs, and worse it doesn't work for large sparse matrices as it requires O(m*n) storage.

I will probably create a WIP draft PR soon. But I would rather suggest to start with replacing outdated random choice with GPU and memory-friendly alternative

**Comment by samaid** [comment]
For random choice replacement I find the most suitable option (memory and GPU friendly) is VariablePhilox described in this [paper](https://arxiv.org/abs/2106.06161). What do you think?
